src/main.cpp

```diff
@@ -6941,73 +6941,56 @@ bool SendMessages(CNode *pto)
-            }
-
-            if (1)
+            // Send message INV up to the MAX_INV_ELEMENTS. Once we reach the max then send the INV message
+            // and if there is any remaining it will be sent on the next iteration until vInventoryToSend is empty.
+            int nToErase = 0;
             {
                 // BU - here we only want to forward message inventory if our peer has actually been requesting
                 // useful data or giving us useful data.  We give them 2 minutes to be useful but then choke off
                 // their inventory.  This prevents fake peers from connecting and listening to our inventory
                 // while providing no value to the network.
                 // However we will still send them block inventory in the case they are a pruned node or wallet
                 // waiting for block announcements, therefore we have to check each inv in pto->vInventoryToSend.
-                bool chokeTxInv = (pto->nActivityBytes == 0 && (nNow / 1000000 - pto->nTimeConnected) > 120);
-                LOCK(pto->cs_inventory);
+                bool fChokeTxInv = (pto->nActivityBytes == 0 && (nNow / 1000000 - pto->nTimeConnected) > 120);
 
-                int invsz = pto->vInventoryToSend.size();
+                // Make copy of vInventoryToSend while cs_inventory is locked
+                const int MAX_INV_ELEMENTS = 1000;
+                int invsz = std::min((int)pto->vInventoryToSend.size(), MAX_INV_ELEMENTS);
                 vInvSend.reserve(invsz);
-                // about 3/4 of the nodes should end up in this list, so over-allocate by 1/10th + 10 items
-                vInvWait.reserve((invsz * 3) / 4 + invsz / 10 + 10);
 
-                // Make copy of vInventoryToSend while cs_inventory is locked but also ignore some tx and defer others
+                LOCK(pto->cs_inventory);
                 for (const CInv &inv : pto->vInventoryToSend)
                 {
+                    nToErase++;
+
                     if (inv.type == MSG_TX)
                     {
-                        if (chokeTxInv)
+                        if (fChokeTxInv)
                             continue;
-                        // skip if we already know abt this one
+                        // skip if we already know about this one
                         if (pto->filterInventoryKnown.contains(inv.hash))
                             continue;
-                        if (fSendTrickle)
-                        {
-                            // 1/4 of tx invs blast to all immediately
-                            if ((insecure_rand.rand32() & 3) != 0)
-                            {
-                                vInvWait.push_back(inv);
-                                continue;
-                            }
-                        }
                     }
                     vInvSend.push_back(inv);
                     pto->filterInventoryKnown.insert(inv.hash);
+
```

Comment:
since we have already the `cs_inventory` lock here, why not removing the element from `vInventoryToSend` by moving to a different for loop structure, something like:

```c++
for (auto inv = vInventoryToSend.begin(); inv != vInventoryToSend.end(); /* do nothing */)  
```

then increase the `inv` iterator only on `continue` on all other case just remove the current `vInventoryToSend` element once we got it added to `vInvSend` and `pto->filterInventoryKnown`